### ISSUE TITLE
Validation block should allow validation of list of values

### ISSUE THREAD

**Issue by avarmaavarma** [hashicorp/terraform#24223]
This validation block works for a single input variable.

```
variable "mytestname" {

     validation {
        condition = length(regexall("^test", var.mytestname)) > 0
        error_message = "Should start with 'test'"
     }
}
```
I need it to work inside a for_each - or have some workaround to accomplish this. The issue is that there is a restriction on the condition statement - the condition HAS to take in the input variable itself (i.e. - it cannot accept an each.value)

```
```
variable "mytestnames" {

listnames = split(",",var.mytestnames)     

for_each = var.listnames

     validation {
        condition = length(regexall("^test", each.value)) > 0
        error_message = "Should start with test"
      }
}

```
The above snippet does not work. The validation condition only accepts the single input variable (var.mylist[0]) etc.  - it does not work inside a loop or allow any element of a collection to be passed in.

**Comment by aaronsteers** [comment]
@avarmaavarma - I believe my feature proposal in #15469 would meet these requirements, essentially adding a `raise()` function which would be able to operate on any combination of variables in a `locals` block.

Would this meet your need?

```
variable "listnames" {
  type = list(string)
}
locals {
    my_test_results = ([
      for v in var.listnames :
      length(regexall("^test", v.value)) > 0 ? "ok" : raise("Should start with 'test'")
    ])
}
```

**Comment by avarmaavarma** [comment]
Yes - that should work for what I was looking for. Also, is there a limit on the number of raises...? Can I raise more than once (provided I do so in a separate block?)

**Comment by aaronsteers** [comment]
@avarmaavarma - Thanks for confirming. This is still just a proposal - but in terms of number of raises, I can't say for sure and it would likely depend on the implementation. That said, at least in theory, the same rules of error handling could/would apply as with the rest of terraform, specifically that multiple errors are reported when not directly dependent upon each other. As the dependency DAG is resolved, obviously anything downstream from the validation check would not get executed.

So I would expect that this code would produce two errors:

```hcl
locals {
  my_int    = "asdf"
  my_calc_a = try(my_int * 5, raise("Can't perform calc A. Is the int really an int?")
  my_calc_b = try(my_int / 5, raise("Can't perform calc B. Is the int really an int?")
}
```

While this would produce a single error:

```hcl
locals {
  my_int    = "asdf"
  my_calc_a = try(my_int * 5, raise("Can't perform calc A. Is the int really an int?")
  my_calc_c = try(my_calc_a * 5, raise("Can't perform calc C. Is the int really an int?")
}
```

Because Calc C depends on an already failed value ("Calc A"), there's no way to execute it and it would not report an error.

**Comment by kappmeier** [comment]
It is a bit clumsy, but using for expression should work:

`length([for x in var.inputs : 1 if contains(["allowed1", "allowed2"], r)]) == length(var.inputs)`

Of course more complex validations are possible.

**Comment by aaronsteers** [comment]
The parallel effort I am covering in the `raise(error_msg)` function PR ( work in progress here: #25088 ) would also make this type of validation possible under a conditional expression living in any `locals` block. If you want to vote for that PR and/or contribute code, this would be much appreciated.

**Comment by aaronsteers** [comment]
Potentially resolved by: #25088 (PR waiting feedback, CI tests passing)

**Comment by sticky-note** [comment]
I've found a quite nice one liner Workaround to validate `list(string)` against another `list(string)`:
```
variable items {
  type = list(string)
  default = []
  validation {
    condition = contains([for item in var.items: contains( ["val1", "val2", "val3"], item)], false)

    error_message = "The name value is empty or contains non-valid value(s)."
  }
}
```
The `condition` test if each element of `items` is one of `["val1", "val2", "val3"]` and emptyness of list as well

Seems to be easily adaptable to your case as:
```
variable "mytestname" {
  type = "string"
  default = ""
  validation {
    condition = contains([for item in split(",",var.mytestnames): length(regexall("^test", item)) > 0], false)
    error_message = "The mytestname must not be empty and items values should start with 'test'"
  }
}
```
Not tested that way and interested to listen if this worked for you.
